File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.js
const fs = require('fs');

const [a, b] = fs.readFileSync('dev/stdin').toString().trim().split("\n");
const [_, s] =a.split(" ").map(Number);
const arr = b.split(" ").map(Number);

const main = () => {
    const li = []
    let cnt = 0
    const recur = (sumv) => {
        if (li.length != 0 && sumv == s){
            cnt += 1;
        }
        
        if (li.length === arr.length) return;
        
        for (let i = 0; i < arr.length; i ++ ){
            if(li.length == 0 || i > li.at(-1)){
                li.push(i)
                recur(sumv + arr[i]);
                li.pop()
            }
        }
    }
    
    recur(0);
    
    console.log(cnt)
}

main()

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.py
import sys

input = sys.stdin.readline
_, s = map(int, input().split())
arr = [int(i) for i in input().split()]

def combi(n, k):
    li = []
    cnt = 0
    
    def recur(sumv):
        nonlocal cnt
        if li and sumv == s:
            cnt += 1
        if len(li) == k:
            return;
        
        for i in range(n):
            if not li or i > li[-1]:
                li.append(i)
                recur(sumv + arr[i])
                li.pop()
                
    recur(0)
    return cnt

def main():
    cnt = combi(len(arr), len(arr))
    return cnt

print(main())
    